// couchdb/couchdb.js
var cradle = require('cradle'),
    _      = require('underscore');

var connection = new(cradle.Connection)('http://home.oszko.net', 5984, {
    auth: {
        username: 'web',
        password: 'web'
    }
});
var dataBuffer = {value:0};
var db = connection.database('jobdata');
//var log = _.bind(console.log, console);
//db.info();
// dbChanges = {
// 	check: function(cb) {
// 			db.changes({since:237}).on('response', function(res) {
// 				res.on('data',function(change) {
// 					//_.delay(cb,5000,true);
// 					console.log(change);
// 									 //cb(true);
// 				})
// 			});
// 		},
// 	ok:false
// };

// dbChanges.check(function(nogut) {
// 	console.log(nogut);
// });
// 
// db.changes().on('response', function(res) {
//  console.log(res);
// });

exports.getSearchData = function(suggest,cb) {
	//var suggestData = {}; 
	// var searchbuffer = {};
	// function sBuffer(data) {
	// 	// body...
	// 	if (!searchbuffer[data]) {
	// 		console.log("not in buffer");
	// 	return searchbuffer[data] = data;
	// 	} else {
	// 		console.log("in buffer");
	// 		return searchbuffer[data];
	// 	}
	// };
	// 
	// console.log(sBuffer(suggest));
	// 
	// return;
	//   if (dataBuffer.value == 0) { 
	// suggestList(function(data) {
	// 	dataBuffer = data;
	// 	cb(dataBuffer);
	// });
	//   } else {
	// 	cb(dataBuffer);
	// 	
	//   }
	
	console.time("getSearchData CouchDB");
	db.view('Jobdata/by_suggestone', {
	    key: suggest,
		  group: true,
	}, function(err, res) {
	    if (err) throw err.error
			//console.dir(res);
			//break
			//suggestData = res;
			//console.dir(suggestData);
			if (_.isEmpty(res)) return cb({noData:true});
			var dataRes = {};
			//console.dir(res[0]);
			//for (i in res) {
			dataRes["couchid"]=_.flatten(res[0].value);
		  //}
			//console.dir(dataRes);
			//return
			// res = {
			// 	
			// 	key: res[0].key,
			// 	value:data
			// };
			//console.dir(res);
			console.timeEnd("getSearchData CouchDB");
			return cb(dataRes);
			});
			

};


exports.checkList = function(cb) {
	console.log("checkList");
	if (dataBuffer.value == 0) { 
	suggestList(function(data) {
		dataBuffer = data;
		cb(dataBuffer);
	});
  } else {
		cb(dataBuffer);
		
  }
	
};


suggestList = function(cb) {
	// body...
	//console.log("function getSuggestList");
	db.view('Jobdata/by_suggesttest', {
	    group: true,
	}, function(err, res) {
	    if (err) throw err.error
			// console.dir(res);
			//suggestData = res;
			//console.dir(suggestData);
			if (_.isEmpty(res)) return cb({noData:true});
			console.log("data get");
			var dataRes = {};
				
			for (i in res) {
				dataRes[res[i].key]=res[i].value;
		  };
			//console.dir(dataRes);
			// res = {
			// 	
			// 	key: res[0].key,
			// 	value:data
			// };
			//console.dir(res);
			//console.log(this);
			
			return cb(dataRes);
	});	
};

exports.suggest = function(suggest,cb) {
	//var suggestData = {}; 
	console.time("jobdata CouchDB");
	db.view('Jobdata/by_suggest', {
	    startkey: suggest,
	    endkey: suggest+"ZZZZ",
	    group: true,
		
	}, function(err, res) {
	    if (err) throw err.error
			// console.dir(res);
			//suggestData = res;
			//console.dir(suggestData);
			if (_.isEmpty(res)) return cb({noData:true});
			var dataRes = {};
			//console.dir(res);
			for (i in res) {
				dataRes[res[i].key]=_.flatten(res[i].value);
		  }
			//console.dir(dataRes);
			// res = {
			// 	
			// 	key: res[0].key,
			// 	value:data
			// };
			//console.dir(res);
			console.timeEnd("jobdata CouchDB");
			return cb(dataRes);
			});
			

};
